lte/cloud/go/services/subscriberdb/storage/sub_proto.go

```diff
@@ -0,0 +1,335 @@
+		)
+		_, err := tx.Exec(deleteRowsOnlyInRealTableQuery)
+		if err != nil {
+			return nil, errors.Wrapf(err, "clean up previous sub proto store table")
+		}
+
+		// the upsertColumnPrefix is used to refer to the the table whose column value is
+		// used in the "on conflict update" operation; it depends on the sql dialect
+		// e.g. for mariaDB, "f1=t2.f1"; for postgres, "f1=excluded.f1"
+		var upsertColumnPrefix string
+		switch getSqlDialect() {
+		case sqorc.PostgresDialect:
+			upsertColumnPrefix = postgresUpsertColumnPrefix
+		case sqorc.MariaDialect:
+			upsertColumnPrefix = subProtoTmpTableName
+		}
+		_, err = l.builder.
+			Insert(subProtoTableName).
+			Select(
+				l.builder.
+					Select(subProtoNidCol, subProtoSidCol, subProtoProtoCol).
+					From(subProtoTmpTableName).
+					Where(squirrel.Eq{subProtoNidCol: network}),
+			).
+			OnConflict(
+				[]sqorc.UpsertValue{{
+					Column: subProtoProtoCol,
+					Value:  squirrel.Expr(fmt.Sprintf("%s.%s", upsertColumnPrefix, subProtoProtoCol)),
+				}},
+				subProtoNidCol, subProtoSidCol,
+			).
+			RunWith(tx).
+			Exec()
+		if err != nil {
+			return nil, errors.Wrapf(err, "populate sub proto store table")
+		}
+
+		_, err = l.builder.
+			Delete(subProtoTmpTableName).
+			Where(squirrel.Eq{subProtoNidCol: network}).
+			RunWith(tx).
+			Exec()
+		if err != nil {
+			return nil, errors.Wrapf(err, "clean up tmp sub proto store table")
+		}
+		return nil, nil
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+func (l *SubProtoStore) ClearTmpTable() error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.
+			Delete(subProtoTmpTableName).
+			RunWith(tx).
+			Exec()
+		return nil, errors.Wrapf(err, "clear sub protos tmp table")
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+// DeleteSubProtos deletes the cached protos for a list of networks.
+func (l *SubProtoStore) DeleteSubProtos(networks []string) error {
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		_, err := l.builder.
+			Delete(subProtoTableName).
+			Where(squirrel.Eq{subProtoNidCol: networks}).
+			RunWith(tx).
+			Exec()
+		return nil, errors.Wrapf(err, "delete sub protos for networks %+v", networks)
+	}
+	_, err := sqorc.ExecInTx(l.db, nil, nil, txFn)
+	return err
+}
+
+// GetPage gets a page of subscriber protos based on the page token and size, and also returns
+// the token for the next page.
+func (l *SubProtoStore) GetPage(network string, token string, pageSize uint64) ([]*lte_protos.SubscriberData, string, error) {
+	lastIncludedSid := ""
+	if token != "" {
+		decoded, err := decodePageToken(token)
+		if err != nil {
+			return nil, "", err
+		}
+		lastIncludedSid = decoded.LastIncludedEntity
+	}
+	txFn := func(tx *sql.Tx) (interface{}, error) {
+		rows, err := l.builder.
+			Select(subProtoProtoCol).
+			From(subProtoTableName).
+			Where(squirrel.And{
+				squirrel.Eq{subProtoNidCol: network},
+				squirrel.Gt{subProtoSidCol: lastIncludedSid},
```

Comment:
Interesting that empty string is less than non-empty string, but makes logic work